src/cq.rs

```diff
@@ -33,6 +41,65 @@ impl CompletionQueueHandle {
     pub fn new() -> CompletionQueueHandle {
         CompletionQueueHandle {
             cq: unsafe { grpc_sys::grpc_completion_queue_create_for_next(ptr::null_mut()) },
+            ref_cnt: AtomicIsize::new(1),
+        }
+    }
+
+    fn add_ref(&self) -> Result<()> {
+        loop {
+            let cnt = self.ref_cnt.load(Ordering::SeqCst);
+            if cnt <= 0 {
+                return Err(Error::QueueShutdown);
+            }
+            let new_cnt = cnt + 1;
+            if cnt ==
+                self.ref_cnt
+                    .compare_and_swap(cnt, new_cnt, Ordering::SeqCst)
+            {
+                return Ok(());
+            }
+        }
+    }
+
+    fn unref(&self) {
+        let shutdown = loop {
+            let cnt = self.ref_cnt.load(Ordering::SeqCst);
+            if cnt == 0 {
+                return;
+            }
+            let new_cnt = cnt - cnt.signum();
```

Comment:
if cnt is -1, new cnt is 0?

Reply:
Yes.

Reply:
can you explain why we need to handle negative here? can we break the loop until the cnt is -1?

Reply:
I think this is explained in https://github.com/pingcap/grpc-rs/pull/125/files#diff-c555e9a56a6bc20f38bbef46ccdd1efbR30.

Reply:
IMO here `cnt - cnt.signum` and later `-cnt + 1` in the shutdown confuse me, and I think it is hard to understand them easily at first glance. 
Why not using a SpinLock to control the reference counter and shutdown, or using two atomic value for counter and shutdown.

Reply:
As I said in https://github.com/pingcap/grpc-rs/pull/125#discussion_r155700879, `SpinLock` is not appropriate here. The usages of atomic value here are just very basic. They are just like `fetch_add` or any other encapsulated functions except that they don't get fancy names.

Only using one atomic value here to get better performance and smaller struct size. And I think the code here is also very straightforward and also commented.

Reply:
As I said before, when I first see `cnt - cnt.signum()` and `-cnt + 1`, I have to think deeply what is the meaning of this, it is not straightforward. Can you add some comments?